Update enhancement for aws-load-balancer-operator (#1107)
openshift/enhancements: enhancements/ingress/aws-load-balancer-operator.md

The operator only reconciles a single instance of the custom resource named `"cluster"` because
only [one
instance](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2185)
of the lb-controller can be run in a cluster. There can be multiple replicas with the same
configuration with leader-election enabled. But multiple controllers cannot be started with
different configurations.

The lb-controller also requires that the subnets where the load balancers are
provisioned have certain resource tags present on them. The operator can detect
the subnets and tag them or the user can do this. This can be enabled by setting
the `SubnetTagging` to `AutoSubnetTaggingPolicy`.

The user can also specify the Ingress class which the lb-controller will
reconcile. This defaults to **_“alb”_**. This value is required because if
an ingress class value is not passed to the controller it attempts to reconcile
all Ingress resources which don’t have an ingress class annotation or where the
ingress class field is empty. In order to prevent this, the operator creates an
**IngressClass** if it cannot find any **IngressClass**  with the name mentioned

> in `.Spec.IngressClass` field. Passing the value of **_“alb”_** does not restrict

Reviewer: It seems this change was also applied and then reverted in the latest push.  
```suggestion
in the `.Spec.IngressClass` field of the CR. Passing the value of **_“alb”_** does not restrict
```

the number of ingress classes which the controller can reconcile. This is

> described further in the [Parallel operation of OpenShift Router](#parallel-operation-of-the-openshift-router-and-lb-controller)
> and lb-controller section.

Reviewer: This change also got reverted.  
```suggestion
described further in the [Parallel operation of the OpenShift Router and lb-controller](#parallel-operation-of-the-openshift-router-and-lb-controller)
section.
```

### Risks and Mitigations

The annotation `alb.ingress.kubernetes.io/target-type` can be set to the
value **_ip_** or **_instance_**. When the value is set to _instance_ (default)
the created load balancer will be configured to route traffic through a NodePort
service. Setting the annotation to _ip_ is not huge risk, since it only works when
the cluster uses the Amazon VPC CNI and hence will not work on OCP/OKD clusters.

### Drawbacks

Since we are reusing an existing upstream controller and restricting which of its
features are enabled, some upstream documentation would not be applicable for this
operator.

The _lb-controller_ currently has support for _Service_ type resources as well.
When the annotation `service.beta.kubernetes.io/aws-load-balancer-type: "external"`
is specified on a _Service_ resource of type _LoadBalancer_ the in-tree controller
ignores this resource and the lb-controller instead provisions a Network Load
Balancer with the correct configuration.

The operator utilises the `service-ca-controller` to generate and inject certificates
using the associated secret. Upon renewal of these certificates, the operator and the

> controller reloads the updated certificates. 

Author: Fixed

balancers.

The operator will also create a
[CredentialRequest](https://docs.openshift.com/container-platform/4.10/rest_api/security_apis/credentialsrequest-cloudcredential-openshift-io-v1.html#credentialsrequest-cloudcredential-openshift-io-v1)
on behalf of the lb-controller and mount the minted credentials in the
_Deployment_ of the controller. This means that the operator will only work on
OCP/OKD but this is sufficient for the initial release. There is a variation in
this flow when it comes to OCP/OKD clusters running in STS or Manual Mode. Since
the `cloud-credentials-operator` will be running in `manual mode`, the credentials
for the lb-controller won't be provisioned automatically, and will require manual
intervention to do so. The required manual steps will be documented within the

> operator docs. Additional documentation for STS mode/manual mode can be found [here](https://docs.openshift.com/container-platform/4.10/authentication/managing_cloud_provider_credentials/cco-mode-sts.html)

Reviewer: This also got reverted.  
```suggestion
operator docs. Additional documentation for STS mode/manual mode can be found in [the OCP documentation for using manual mode with STS](https://docs.openshift.com/container-platform/4.10/authentication/managing_cloud_provider_credentials/cco-mode-sts.html)
```

The lb-controller requires a validating and mutating webhook for correct operation.

> The operator will have to create the webhooks along with the controller deployment. 

Reviewer: This also got reverted.  
```suggestion
The operator creates the webhook configuration along with the controller deployment. 
```

The webhook can be registered with a CA bundle which is used to verify the identity 
of webhook by the API server. The [service-ca controller](https://docs.openshift.com/container-platform/4.10/security/certificate_types_descriptions/service-ca-certificates.html)
can be used to generate certificates and have them injected into the webhook
configurations. The operator does not manage the lifecycle of the certificates
in the associated secret. Additional details mentioned in [Drawbacks](#drawbacks).

### Risks and Mitigations

> The annotation `alb.ingress.kubernetes.io/target-type` can be set to the
> value **_ip_** or **_instance_**. When the value is set to _instance_ (default)
> the created load balancer will be configured to route traffic through a NodePort
> service. Setting the annotation to _ip_ is not huge risk, since it only works when
> the cluster uses the Amazon VPC CNI and hence will not work on OCP/OKD clusters.

Reviewer: I'm confused.  Is the risk that the user might set this annotation to a problematic value?  If the user does set it to "ip", what happens?  (What does "will not work mean"?)  Is this what the `EnableIPTargetType` feature gate was added to prevent?  
```suggestion
The user can set the `alb.ingress.kubernetes.io/target-type` annotation to the
value **_ip_** or **_instance_**. When the value is set to _instance_,
the created load balancer will be configured to route traffic through a NodePort
service.  This is also the default behavior when the annotation is not set.
However, when the value is set to _ip_, the lb-controller requires that
the cluster use the Amazon VPC CNI, and hence [the lb-controller
 will not do anything|the lb-controller will behave the same as if
 "instance" were specified|something else?] on OCP/OKD clusters.
```

Author: If the user does set the value to `ip`, the controller errors since the `TargetGroupBinding` cannot resolve pod ENI since it is dependent on AWS VPC CNI. 
>`
{"level":"error","ts":1654582687.4895403,"logger":"controller-runtime.manager.controller.targetGroupBinding","msg":"Reconciler error","reconciler group":"elbv2.k8s.aws","reconciler kind":"TargetGroupBinding","name":"k8s-awsloadb-echoserv-362ca47f2d","namespace":"aws-load-balancer-test-custom-ing","error":"cannot resolve pod ENI for pods: [aws-load-balancer-test-custom-ing/echoserver]"}
`

Yes, `EnableIPTargetType` was introduced to prevent this. It is yet to be released upstream, hence the current risk.

Reviewer: `EnableIPTargetType` is carried in the downstream fork though, right?  Anyway, can you update the text to clarify the issues raised in my earlier comment?

### Drawbacks

Since we are reusing an existing upstream controller and restricting which of its
features are enabled, some upstream documentation would not be applicable for this
operator.

The _lb-controller_ currently has support for _Service_ type resources as well.
When the annotation `service.beta.kubernetes.io/aws-load-balancer-type: "external"`
is specified on a _Service_ resource of type _LoadBalancer_ the in-tree controller
ignores this resource and the lb-controller instead provisions a Network Load
Balancer with the correct configuration.

The operator utilises the `service-ca-controller` to generate and inject certificates
using the associated secret. Currently upon renewal of these certificates, the operator
does not reload the deployment. Dynamic reload of the deployment isn't supported in

> the initial release.

Reviewer: This means that the webhook will fail and break your cluster if the certificate expires, right?  So that's a 2-year timebomb.  It's crucial we not lose track of that.

Author: Noted, will have a corresponding task in the backlog.

Author: This issue was already mitigated in the initial release. The controller uses webhook rotation [feature](https://github.com/kubernetes-sigs/controller-runtime/pull/421) provided by the container runtime. 

Also verified manually, 
```bash
{"level":"info","ts":1655811187.8988125,"logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
{"level":"info","ts":1655811187.898957,"logger":"controller-runtime.certwatcher","msg":"Updated current TLS certificate"}
```

TBD

## Implementation History

- Update enhancement for aws-load-balancer-operator [#1107](https://github.com/openshift/enhancements/pull/1107)
  - Updates some implementation details
  - Updates test plan and current caveats in testing

>   - Removes non-applicable details for initial release

Author: Fixed by removing revision details.
